# Exploratory Data Analysis

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import plotly.express as px

df = pd.read_csv(
    "../data/trustmesh_economic_data.csv",
    parse_dates=["month"],
    keep_default_na=False
)

print(f"Dataset shape: {df.shape}")
print(f"Duplicate rows: {df.duplicated().sum()}")
print(f"Missing values: {df.isna().sum().sum()}")

df.head()

Dataset shape: (12000, 32)
Duplicate rows: 0
Missing values: 0


,business_id,owner_id,business_type,environment_type,location,month,monthly_income,monthly_expenses,transaction_count,digital_payment_pct,...,shock_factor,adjusted_income,adjusted_expenses,adjusted_cash_flow,cash_flow_margin_pct,expense_burden_pct,operational_stability_pct,financial_health_pct,resilience_pct,economic_reputation_index
0,B001,O001,Small Producer,Rural,Patna,2024-01-01,28232.13,20097.60,61,36.62,...,1.0,26820.52,20097.60,1448.69,5.40,74.93,74.68,27.66,75.96,57.71
1,B001,O001,Small Producer,Rural,Patna,2024-02-01,30774.11,19325.85,61,23.46,...,1.0,29850.89,19325.85,4192.18,14.04,64.74,80.10,34.64,83.14,62.98
2,B001,O001,Small Producer,Rural,Patna,2024-03-01,31252.98,20911.70,77,43.26,...,1.0,31252.98,20911.70,4456.40,14.26,66.91,72.56,35.16,71.67,59.83
3,B001,O001,Small Producer,Rural,Patna,2024-04-01,37389.18,23891.42,55,46.95,...,1.0,38136.96,23891.42,7681.90,20.14,62.65,78.88,40.53,78.15,65.75
4,B001,O001,Small Producer,Rural,Patna,2024-05-01,34419.97,20059.14,74,36.71,...,1.0,36140.97,20059.14,10302.91,28.51,55.50,77.06,44.81,77.90,65.44


In [ ]:

print("Shape:", df.shape)
print("\nBusiness types:")
print(df["business_type"].value_counts())

print("\nEnvironment types:")
print(df["environment_type"].value_counts())

print("\nLocations:")
print(df["location"].value_counts())

print("\nDate range:")
print(df["month"].min(), "to", df["month"].max())

df.describe().T

Shape: (12000, 32)

Business types:
business_type
Retail Shop               2400
Food Vendor               1920
Small Producer            1440
Transport Worker          1440
Micro Service Business    1200
Home Business             1200
Freelancer                1200
Artisan                   1200
Name: count, dtype: int64

Environment types:
environment_type
Urban         4200
Rural         3000
Semi-Urban    3000
Peri-Urban    1800
Name: count, dtype: int64

Locations:
location
Delhi        1416
Guwahati     1320
Bengaluru    1296
Jaipur       1224
Indore       1224
Patna        1176
Mumbai       1176
Kochi        1104
Lucknow      1032
Ahmedabad    1032
Name: count, dtype: int64

Date range:
2024-01-01 00:00:00 to 2025-12-01 00:00:00


,count,mean,min,25%,50%,75%,max,std
month,12000,2024-12-15 22:00:00,2024-01-01 00:00:00,2024-06-23 12:00:00,2024-12-16 12:00:00,2025-06-08 12:00:00,2025-12-01 00:00:00,NaN
monthly_income,12000.0,34001.574249,16812.47,28337.965,32278.355,39229.0175,60948.29,7533.094634
monthly_expenses,12000.0,20699.510107,8214.17,15597.135,18696.72,24271.0475,45463.17,6975.864655
transaction_count,12000.0,68.008167,19.0,56.0,66.0,79.0,132.0,17.044378
digital_payment_pct,12000.0,54.386884,8.86,42.5375,53.12,64.48,100.0,16.292415
supplier_payment_amount,12000.0,6795.206652,1027.38,2813.465,5840.995,9636.2375,22262.61,4688.828763
utility_payment_amount,12000.0,1348.130212,292.6,652.5275,1170.18,1908.405,3976.42,791.944714
rent_payment,12000.0,6588.705338,2747.29,4661.1375,6392.455,8544.36,11728.78,2035.92893
savings_amount,12000.0,1011.30961,0.0,439.8675,808.65,1335.83,6194.57,838.938975
net_cash_flow,12000.0,5365.228592,-5731.5,2243.18,4527.765,6987.675,26159.15,4960.008238


In [ ]:

fig = px.bar(
    df["business_type"].value_counts().reset_index(),
    x="business_type",
    y="count",
    title="Business Type Distribution",
    labels={"business_type": "Business Type", "count": "Number of Observations"}
)

fig.show()

In [5]:
df["location"].value_counts()

location
Delhi        1416
Guwahati     1320
Bengaluru    1296
Jaipur       1224
Indore       1224
Patna        1176
Mumbai       1176
Kochi        1104
Lucknow      1032
Ahmedabad    1032
Name: count, dtype: int64

In [ ]:


financial_by_type = (
    df.groupby("business_type")
      .agg(
          avg_income=("adjusted_income", "mean"),
          avg_expenses=("adjusted_expenses", "mean"),
          avg_cash_flow=("adjusted_cash_flow", "mean"),
          avg_transactions=("transaction_count", "mean"),
          avg_digital_payment=("digital_payment_pct", "mean"),
          avg_expense_burden=("expense_burden_pct", "mean")
      )
      .round(2)
      .sort_values("avg_cash_flow", ascending=False)
)

display(financial_by_type)


for metric, title, y_label in [
    ("avg_income", "Average Adjusted Income by Business Type", "Average Income"),
    ("avg_cash_flow", "Average Adjusted Cash Flow by Business Type", "Average Cash Flow"),
    ("avg_expense_burden", "Average Expense Burden by Business Type", "Expense Burden (%)")
]:
    fig = px.bar(
        financial_by_type.reset_index(),
        x="business_type",
        y=metric,
        title=title,
        labels={
            "business_type": "Business Type",
            metric: y_label
        }
    )
    fig.show()

,avg_income,avg_expenses,avg_cash_flow,avg_transactions,avg_digital_payment,avg_expense_burden
business_type,,,,,,
Freelancer,40682.76,16332.27,17513.26,80.15,84.99,40.64
Micro Service Business,30223.74,16718.18,6146.93,59.60,59.80,55.98
Transport Worker,28286.04,15673.50,5442.21,56.11,49.87,56.15
Home Business,25261.84,12714.78,5415.53,50.01,55.39,50.95
Artisan,32377.45,19579.51,4951.68,64.23,40.29,61.29
Retail Shop,45350.68,32859.71,3424.82,89.82,64.97,73.42
Small Producer,35189.44,24045.64,2632.24,69.42,34.69,69.24
Food Vendor,30280.32,19793.92,2208.82,59.89,45.00,66.16


In [ ]:


reputation_by_type = (
    df.groupby("business_type")
      .agg(
          financial_health=("financial_health_pct", "mean"),
          operational_stability=("operational_stability_pct", "mean"),
          resilience=("resilience_pct", "mean"),
          reputation_index=("economic_reputation_index", "mean")
      )
      .round(2)
      .sort_values("reputation_index", ascending=False)
)

display(reputation_by_type)


fig = px.bar(
    reputation_by_type.reset_index(),
    x="business_type",
    y="reputation_index",
    title="Average Economic Reputation Index by Business Type",
    labels={
        "business_type": "Business Type",
        "reputation_index": "Economic Reputation Index"
    }
)

fig.show()


reputation_long = (
    reputation_by_type
    .reset_index()
    .melt(
        id_vars="business_type",
        var_name="metric",
        value_name="score"
    )
)

fig = px.bar(
    reputation_long,
    x="business_type",
    y="score",
    color="metric",
    barmode="group",
    title="Reputation Dimensions by Business Type",
    labels={
        "business_type": "Business Type",
        "score": "Score",
        "metric": "Dimension"
    }
)

fig.show()

,financial_health,operational_stability,resilience,reputation_index
business_type,,,,
Freelancer,58.02,84.31,84.13,75.18
Micro Service Business,42.85,82.43,82.36,68.60
Home Business,44.41,79.25,78.66,67.29
Retail Shop,32.75,81.39,81.22,64.44
Artisan,38.58,77.65,76.88,64.28
Transport Worker,40.34,73.85,73.57,62.23
Food Vendor,33.77,76.67,76.36,61.78
Small Producer,32.42,75.49,75.32,60.49


In [ ]:


shock_analysis = (
    df.groupby("shock_type")
      .agg(
          observations=("business_id", "size"),
          avg_income=("adjusted_income", "mean"),
          avg_expenses=("adjusted_expenses", "mean"),
          avg_cash_flow=("adjusted_cash_flow", "mean"),
          avg_financial_health=("financial_health_pct", "mean"),
          avg_resilience=("resilience_pct", "mean"),
          avg_reputation=("economic_reputation_index", "mean")
      )
      .round(2)
      .sort_values("avg_reputation", ascending=False)
)

display(shock_analysis)


fig = px.bar(
    shock_analysis.reset_index(),
    x="shock_type",
    y="avg_cash_flow",
    title="Average Adjusted Cash Flow by Economic Shock",
    labels={
        "shock_type": "Economic Shock",
        "avg_cash_flow": "Average Adjusted Cash Flow"
    }
)

fig.show()


fig = px.bar(
    shock_analysis.reset_index(),
    x="shock_type",
    y="avg_reputation",
    title="Average Economic Reputation by Economic Shock",
    labels={
        "shock_type": "Economic Shock",
        "avg_reputation": "Economic Reputation Index"
    }
)

fig.show()

,observations,avg_income,avg_expenses,avg_cash_flow,avg_financial_health,avg_resilience,avg_reputation
shock_type,,,,,,,
None,9818,35517.48,20705.34,6881.34,41.24,78.58,65.82
Supply Disruption,713,30429.71,22871.59,-445.97,30.28,78.57,61.92
Operating Cost Increase,641,31395.48,24339.52,-806.38,29.34,78.04,61.42
Demand Drop,828,26700.53,20872.90,-2180.58,28.46,78.24,61.13


In [ ]:


monthly_trends = (
    df.groupby("month")
      .agg(
          avg_income=("adjusted_income", "mean"),
          avg_cash_flow=("adjusted_cash_flow", "mean"),
          avg_reputation=("economic_reputation_index", "mean")
      )
      .reset_index()
)

display(monthly_trends.head())
fig = px.line(
    monthly_trends,
    x="month",
    y="avg_income",
    title="Monthly Average Adjusted Income",
    labels={
        "month": "Month",
        "avg_income": "Average Adjusted Income"
    }
)
fig.show()


fig = px.line(
    monthly_trends,
    x="month",
    y="avg_cash_flow",
    title="Monthly Average Adjusted Cash Flow",
    labels={
        "month": "Month",
        "avg_cash_flow": "Average Adjusted Cash Flow"
    }
)
fig.show()


fig = px.line(
    monthly_trends,
    x="month",
    y="avg_reputation",
    title="Monthly Average Economic Reputation",
    labels={
        "month": "Month",
        "avg_reputation": "Economic Reputation Index"
    }
)
fig.show()

,month,avg_income,avg_cash_flow,avg_reputation
0,2024-01-01,31203.51134,2259.04878,63.63388
1,2024-02-01,31750.37906,2903.77722,63.87508
2,2024-03-01,33062.56550,3991.89454,64.43264
3,2024-04-01,33356.83422,4486.24366,64.81786
4,2024-05-01,34484.00964,5721.14598,65.18854


In [ ]:


numeric_cols = df.select_dtypes(include="number").columns

reputation_corr = (
    df[numeric_cols]
    .corr()["economic_reputation_index"]
    .drop("economic_reputation_index")
    .sort_values(ascending=False)
)

display(reputation_corr.to_frame("correlation"))

corr_cols = [
    "adjusted_income",
    "adjusted_expenses",
    "adjusted_cash_flow",
    "digital_payment_pct",
    "payment_discipline_pct",
    "supplier_reliability_pct",
    "business_continuity_pct",
    "demand_stability_pct",
    "cash_flow_margin_pct",
    "expense_burden_pct",
    "financial_health_pct",
    "operational_stability_pct",
    "resilience_pct",
    "economic_reputation_index"
]

fig = px.imshow(
    df[corr_cols].corr().round(2),
    text_auto=True,
    title="Correlation Matrix of TrustMesh Economic Indicators",
    aspect="auto"
)

fig.show()

,correlation
financial_health_pct,0.839560
operational_stability_pct,0.762210
adjusted_cash_flow,0.736499
cash_flow_margin_pct,0.723845
net_cash_flow,0.664181
payment_discipline_pct,0.656196
resilience_pct,0.612317
digital_payment_pct,0.547255
savings_amount,0.531074
business_continuity_pct,0.449663


## Key Findings

### 1. Financial Performance Varies Across Business Types

Retail Shops generate the highest average adjusted income at approximately ₹45,351 per month. However, Freelancers generate the highest average adjusted cash flow at approximately ₹17,513, despite having lower average income than Retail Shops.

This difference is largely explained by expense burden. Retail Shops have an average expense burden of approximately 73.42%, compared with only 40.64% for Freelancers.

### 2. Economic Reputation Is Not Determined by Income Alone

Freelancers have the highest average Economic Reputation Index at 75.18, while Retail Shops have a lower average reputation score of 64.44 despite generating the highest average income.

This demonstrates that TrustMesh evaluates broader economic health rather than relying on revenue alone.

### 3. Financial Health Is the Strongest Reputation Signal

Financial health has the strongest correlation with the Economic Reputation Index (r = 0.84), followed by operational stability (r = 0.76), adjusted cash flow (r = 0.74), and cash-flow margin (r = 0.72).

Expense burden has a strong negative relationship with reputation (r = -0.67).

### 4. Economic Shocks Reduce Financial Performance and Reputation

Businesses experiencing economic shocks show substantially weaker financial outcomes than businesses without shocks.

Demand Drop produces the largest average cash-flow impact, with average adjusted cash flow falling to approximately -₹2,181 compared with ₹6,881 for businesses without a shock.

Demand Drop also produces the lowest average financial health (28.46) and Economic Reputation Index (61.13).

### 5. Reputation Changes Over Time

The monthly analysis shows variation in income, cash flow, and Economic Reputation across the 24-month observation period.

The seasonal factors introduced into the dataset create observable changes in economic performance across different months, while economic shocks introduce additional variation.

### 6. Behavioral Signals Contribute to Economic Reputation

Payment discipline, business continuity, supplier reliability, and demand stability all show positive relationships with the Economic Reputation Index.

This supports the idea that TrustMesh can evaluate businesses using a combination of financial and behavioral signals rather than financial performance alone.

### Overall Observation

The exploratory analysis indicates that economic reputation is better represented as a multidimensional measure of financial health, operational stability, resilience, and business behavior rather than a simple measure of business revenue.